pkg/controllers/networktopologyaware/topoautodiscovery/cache/cache.go

```diff
@@ -0,0 +1,363 @@
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corelister "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	yaml "sigs.k8s.io/yaml/goyaml.v2"
+
+	topologyv1alpha1 "volcano.sh/apis/pkg/apis/topology/v1alpha1"
+	topologylister "volcano.sh/apis/pkg/client/listers/topology/v1alpha1"
+)
+
+const (
+	LabelBasedTopologyName = "volcano.io/label-based-topology-name"
+)
+
+type Cache interface {
+	// OpenSession loads configuration from configmap and compare the current Hypernodes in cluster with th configurtion,
+	// determine which HyperNodes need to be synchronized or deleted.
+	// The resulting HyperNodes are cached
+	OpenSession()
+	// CloseSession clear all caches
+	CloseSession()
+	// GetHyperNodeToGC consumes the gc cache
+	GetHyperNodeToGC() (hyperNode string, done bool)
+	// GetHyperNodeToSync consumes the synced cache
+	GetHyperNodeToSync() (hyperNode *topologyv1alpha1.HyperNode, done bool)
+}
+
+type HyperNodeCache struct {
+	namespace       string
+	configMapLister corelister.ConfigMapLister
+	nodeLister      corelister.NodeLister
+	hyperNodeLister topologylister.HyperNodeLister
+
+	sync.RWMutex
+	sessionOpened            bool
+	topologyConfigDirty      bool
+	nodeLabelBasedTopologies NodeLabelBasedTopologies
+	dirtyHyperNodes          sets.Set[string]
+	desiredHyperNodes        sets.Set[*topologyv1alpha1.HyperNode]
+	hyperNodeToGC            sets.Set[string]
+}
+
+func NewHyperNodeCache(namespace string, nodeInformer coreinformers.NodeInformer, configMapInformer coreinformers.ConfigMapInformer,
+	hyperNodeLister topologylister.HyperNodeLister) Cache {
+	c := &HyperNodeCache{
+		namespace:       namespace,
+		configMapLister: configMapInformer.Lister(),
+		nodeLister:      nodeInformer.Lister(),
+		hyperNodeLister: hyperNodeLister,
+
+		dirtyHyperNodes:   sets.New[string](),
+		desiredHyperNodes: sets.New[*topologyv1alpha1.HyperNode](),
+		hyperNodeToGC:     sets.New[string](),
+	}
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addNode,
+		UpdateFunc: c.updateNode,
+		DeleteFunc: c.deleteNode,
+	})
+
+	configMapInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.updateConfigMap,
+		})
+	return c
+}
+
+func (c *HyperNodeCache) GetNodeLabelBasedTopologies() NodeLabelBasedTopologies {
+	c.RLock()
+	defer c.RUnlock()
+
+	return c.nodeLabelBasedTopologies
+}
+
+func (c *HyperNodeCache) DirtyTopologyConfig() {
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.sessionOpened || c.topologyConfigDirty {
+		return
+	}
+	c.topologyConfigDirty = true
+}
+
+func (c *HyperNodeCache) DirtyHyperNode(name string) {
```

Comment:
I want to make sure that does `DirtyHyperNodes` mean that if during a round of session for constructing the HyperNodes tree, a node is added or there is any change of the node label, this HyperNode cannot be constructed and has to wait until the next round to be built? Is this correct?